docs/rfc/rfc-001-storage-engine.rst

```diff
@@ -0,0 +1,179 @@
+===========================================
+RFC 001: Storage Engines and Database Layer
+===========================================
+
+Changelog
+---------
+
+- 2021-04-19: Initial Draft (gist)
+- 2021-09-02: Migrated to RFC folder, with some updates  
+
+Abstract
+--------
+
+The aspect of Tendermint that's responsible for persistence and storage (often
+"the database" internally) represents a bottle neck in the architecture of the
+platform, that the 0.36 release presents a good opportunity to correct. The
+current storage engine layer provides a great deal of flexibility that is
+difficult for users to leverage or benefit from, while also making it harder
+for Tendermint Core developers to deliver improvements on storage engine. This
+RFC discusses the possible improvements to this layer of the system.
+
+Background
+----------
+
+Tendermint has a very thin common wrapper that makes Tendermint itself
+(largely) agnostic to the data storage layer (within the realm of the popular
+key-value/embedded databases.) This flexibility is not particularly useful:
+the benefits of a specific database engine in the context of Tendermint is not
+particularly well understood, and the maintenance burden for multiple backends
+is not commensurate with the benefit provided. Additionally, because the data
+storage layer is handled generically, and most tests run with an in-memory
+framework, it's difficult to take advantage of any higher-level features of a
+database engine.
+
+Ideally, developers within Tendermint will be able to interact with persisted
+data via an interface that can function, approximately like an object
+store, and this storage interface will be able to accommodate all existing
+persistence workloads (e.g. block storage, local peer management information
+like the "address book", crash-recovery log like the WAL.) In addition to
```

Comment:
I think moving the WAL to use a shared DB would be a great way to get us out of maintaining a homegrown but critical system. I do wonder, however, what amount of updating to the WAL we want to do within this single project? We could definitely reevaluate the WAL and attempt to remove WAL access where it is not necessary for crash-recovery and correctness, but I worry that a large WAL audit will blow up the scope of the DB change out project.  I think it might be more wise for us to use the new DB as a drop-in replacement for the current WAL in this project.

Reply:
This is a good question, but if we _do_ use Badger (or similar) it may obviate the need for a WAL: It provides commit semantics that don't require us to maintain our own persistence log. 

Having said that, I don't know much about how our existing WAL is used, so I think it's still worth addressing.

Reply:
One of the incoming assumptions about this project is that badger's (or similar) would have crash resistance that would obviate the need for the independent WAL. 

The WAL is pretty limited in scope right now, and I would end up doing the audit you propose as part of a more indepth rationalization/modernization of the consensus reactor (which may or may not be obvious.)